st2common/st2common/util/output_schema.py

```diff
@@ -53,22 +54,39 @@ def _validate_action(action_schema, result, output_key):
 
 
 def mask_secret_output(ac_ex, output_value):
-    if not output_value:
-        return output_value
-
-    if not isinstance(output_value, dict) and not isinstance(output_value, list):
+    # We only support output_schema validation when the output_value is a JSON object.
+    # Invididual keys of that object can be marked secret, but the entire output
+    # object cannot be marked as secret.
+    # FIXME: Should we support non-objects under output_key?
+    #        Changing that will require changes in one or more of:
+    #          st2common/util/schema/action_output_schema.json
+    #          st2common/util/schema/__init__.py
+    #          st2common/models/api/action.py
+
+    if not output_value or not isinstance(output_value, Mapping):
```

Comment:
Are we sure list is not supported?  I thought json schema can be applied to a list. Is the output schema for the output value defined correctly?

Reply:
Yes jsonschema can validate all JSON types, but `action_output_schema.json` currently only allows for object.

Reply:
Oh. I see where I misread the schema. Of course the schema itself is an object. And it accepts any of the types:
https://github.com/StackStorm/st2/blob/b8402b4c730a4d20797d84e584a8ba20e6e2684a/st2common/st2common/util/schema/action_output_schema.json#L136-L140

So, yes, output_schema already supports all simple types, so we need to support all of those types as well.

Reply:
`output_schema` != `action_output_schema.json`
https://github.com/StackStorm/st2/blob/b8402b4c730a4d20797d84e584a8ba20e6e2684a/st2common/st2common/models/api/action.py#L231-L237

So, the output_schema begins with the premise of describing a list of properties. Thus the top-level `output_value[output_key]` object has to be an object in order to be validated.

So, `st2common/models/api/action.py` is the place that needs to change if we want the output_schema to describe any other types.